features/features.go

```diff
@@ -309,3 +252,57 @@ func FromContext(ctx context.Context) *Features {
 	}
 	return nil
 }
+
+func (f *Features) Adjust() ([]string, error) {
+	warningStrs := make([]string, 0)
+	
+	v := reflect.ValueOf(f).Elem()
+	t := v.Type()
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		value := v.Field(i)
+		version := field.Tag.Get("enabled_in_version")
+		if len(version) == 0 {
+			continue
+		}
+		majorVersion, minorVersion := ParseVersion(field.Tag.Get("enabled_in_version"))
+		if versionAtLeast(*f, majorVersion, minorVersion) && value.Kind() == reflect.Bool {
+			if value.Bool() {
+				tagName := field.Tag.Get("long")
+				warningStrs = append(warningStrs, fmt.Sprintf("--%s", strings.ToLower(tagName)))
+			}
+		value.SetBool(true)
+		}
+	}
+
+	processNegativeFlags(f)
+
+	if f.ArgScopeSet && !f.ShellOutAnywhere {
+		// ArgScopeSet uses new ARG declaration logic that requires
+		// ShellOutAnywhere. We're erroring here to ensure that users get that
+		// feedback as early as possible.
+		return nil, errors.New("--arg-scope-and-set requires --shell-out-anywhere")
+	}
+
+	return warningStrs, nil
+}
+
+func ParseVersion(version string) (int, int) {
```

Comment:
since this function panics, we should rename it to `must....` to make it more clear.

```suggestion
func MustParseVersion(version string) (int, int) {
```